browser/src/UI/components/StatusResize.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from "react"
+import styled from "styled-components"
+
+import { withProps } from "./common"
+
+interface IChildDimensions {
+    direction: string
+    width: number
+    id: string
+    priority: number
+    hide: boolean
+}
+
+type PassWidth = (data: IChildDimensions) => void
+
+interface Props {
+    children?: React.ReactNode
+    className?: string
+    passWidth?: PassWidth
+    direction: string
+}
+
+interface State {
+    containerWidth: number
+    children: {
+        [id: string]: {
+            id: string;
+            width: number;
+            alignment: number;
+            hide?: boolean;
+            priority: number;
+            passWidth: PassWidth;
+        };
+    }
+}
+
+interface Section {
+    direction: string
+    count: boolean
+}
+
+const StatusbarSection = withProps<Section>(styled.div)`
+    flex: 1 1 auto;
+    display: ${({ count }) => count ? `none` : `flex`};
+    flex-direction: row;
+    height: 100%;
+    max-width: 50%;
+    justify-content: ${props => props.direction};
+`
+
+class StatusBarResizer extends React.Component<Props, State> {
+    private observer: any
+    private elem: Element
+    constructor(props: Props) {
+        super(props)
+        this.state = {
+            containerWidth: null,
+            children: {},
+        }
+    }
+
+    public componentDidMount() {
+        this.setState({
+            containerWidth: this.elem.getBoundingClientRect().width,
+        })
+        // tslint:disable-next-line
+        this.observer = new window["ResizeObserver"](([entry]: any) => {
+            this.setState({ containerWidth: entry.contentRect.width }, this.resize)
+        })
+        this.observer.observe(this.elem)
+        this.resize()
+    }
+
+    public componentWillUnmount() {
+        this.observer.disconnect()
+    }
+
+    public render() {
+        const { containerWidth } = this.state
+        const { children, direction } = this.props
+        const count = React.Children.count(children)
+        return (
+            <StatusbarSection
+                direction={direction}
+                count={count < 1}
+                innerRef={(elem: Element) => (this.elem = elem)}
+            >
+                {containerWidth !== undefined &&
+                    React.Children.map(children, (child: any) => {
+                        const current = this.state.children[child.props.id]
+                        return React.cloneElement(child, {
```

Comment:
I wasn't quite sure why we need to clone the element here?

Reply:
@bryphe cloning the element allows me to add to its properties inside of where its rendered, so that I can clone the element add a method from its parent to the clone as a prop and reference that method in the child, for example inside of the clonedChild aka `statusBarItem` I am able to use `passWidth` a reference to the parents passWidth method or add a hide prop which toggles visibility based on the parent's state/logic. 

It's crucial to this working and as far as I know the way to attach data or functions to react.children given that here we render an unspecified number of children and they all need this behaviour and the only point at which this data is available is here.

Reply:
Got it, I see, so this is used to augment the set of properties. Makes sense - thanks for the explanation!